src/acp/rule.ts

```diff
@@ -0,0 +1,235 @@
+/**
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { acp } from "../constants";
+import { SolidDataset, ThingPersisted, UrlString } from "../interfaces";
+import { addIri } from "../thing/add";
+import { getIriAll } from "../thing/get";
+import { removeAll, removeIri } from "../thing/remove";
+import { setIri } from "../thing/set";
+import { Policy } from "./policy";
+
+export type RuleDataset = SolidDataset;
+export type Rule = ThingPersisted;
```

Comment:
I just realised: when separating Policies from Controls, I made it so that you didn't give the actual _Policies_ to the Controls, but their _URLs_. This to indicate that you have to make sure that the Policies are saved somewhere, not necessarily in the same Resource as the Controls themselves.

I think something similar should apply here. Even though usually a Rule will (hopefully) be located in the same Resource as the Policies, it might not be. So maybe we should update our functions to say `addRequiredRuleUrlToPolicy()` rather than `addRequiredRuleToPolicy()`, just like we have `addPolicyUrl()` rather than `addPolicy()`.

(And I suppose for consistency, the latter should be renamed to `addPolicyUrlToControl()` as well? Though in a separate PR.)

And then maybe this should become:

```suggestion
export type RuleUrl = ThingPersisted | Url | UrlString;
```

Reply:
Firstly, I think I prefer `addRequiredRuleReferenceToPolicy()` over `addRequiredRuleUrlToPolicy()` (so, yeah, that would probably ripple across lots more name re-alignment too, as separate topic/PR), but yeah I agree with the principle of having explicitly named functions to add a Policy/Rule directly, and to add a reference to a Policy/Rule. And yeah, ditto for the Policy and the Access Control Resource.

Reply:
Fixed in eb3d09e413b6b6d9ca69a9df7741411e6d914740